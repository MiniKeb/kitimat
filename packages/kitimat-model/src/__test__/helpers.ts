import { Graph, State, Action, Branch } from '../graph';
import * as K from 'kitimat';

export const noopState = (name: string): State<any, any> => {
  return {
    name: name,
    description: () => `NOOP: ${name}`,
    validate: async (model, oracle) => true,
    branches: [],
  };
};

export const noopAction = (): Action<any, any, any> => {
  return {
    fuzzer: K.constant(undefined),
    description: () => 'noop',
    apply: (model, oracle) => true,
    nextModel: model => model,
    preValidate: () => true,
    postValidate: () => true,
  };
};

export const uniqueByName = (states: State<any, any>[]) =>
  states.reduce(
    (results, next) => {
      if (results.find(val => val.name === next.name)) {
        return results;
      }

      return results.concat(next);
    },
    [] as State<any, any>[],
  );
