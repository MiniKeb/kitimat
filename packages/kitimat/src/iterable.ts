if (!Symbol.asyncIterator) {
  (Symbol as any).asyncIterator = Symbol.for('Symbol.asyncIterator');
}

const getIterator = <A>(iter: AsyncIterable<A>): AsyncIterator<A> => iter[Symbol.asyncIterator]();

export const create = <A>(fn: () => AsyncIterator<A>): AsyncIterable<A> => ({
  [Symbol.asyncIterator]: fn,
});

export const toArray = async <A>(iter: AsyncIterable<A>): Promise<A[]> => {
  const result = [];

  for await (let value of iter) {
    result.push(value);
  }

  return result;
};

export const empty = <A>(): AsyncIterable<A> => create<A>(async function*(): AsyncIterator<A> {});

export const from = <A>(arr: (A | Promise<A>)[]): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* arr;
  });

export const of = <A>(a: A | Promise<A>): AsyncIterable<A> =>
  create<A>(async function*(): AsyncIterator<A> {
    yield await a;
  });

export const every = async <A>(fn: (a: A) => boolean | Promise<boolean>, iter: AsyncIterable<A>): Promise<boolean> => {
  for await (let next of iter) {
    if ((await fn(next)) === false) {
      return false;
    }
  }

  return true;
};

export const some = async <A>(fn: (a: A) => boolean | Promise<boolean>, iter: AsyncIterable<A>): Promise<boolean> => {
  for await (let next of iter) {
    if ((await fn(next)) === true) {
      return true;
    }
  }

  return false;
};

export const lazy = <A>(fn: () => AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* fn();
  });

export const map = <A, B>(fn: (a: A) => B | Promise<B>, iter: AsyncIterable<A>): AsyncIterable<B> =>
  create<B>(async function*() {
    for await (let next of iter) {
      yield await fn(next);
    }
  });

export const map2 = <A, B, C>(
  fn: (a: A, b: B) => C | Promise<C>,
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
): AsyncIterable<C> =>
  create<C>(async function*() {
    const iteratorA = getIterator(a);
    const iteratorB = getIterator(b);

    let resultA: IteratorResult<A>;
    let resultB: IteratorResult<B>;

    do {
      // do this so that they resolve themselves in parallel
      [resultA, resultB] = [await iteratorA.next(), await iteratorB.next()];

      if (resultA.done === false && resultB.done === false) {
        yield await fn(resultA.value, resultB.value);
      }
    } while (resultA.done === false && resultB.done === false);
  });

export const zip = <A, B>(a: AsyncIterable<A>, b: AsyncIterable<B>): AsyncIterable<[A, B]> =>
  map2<A, B, [A, B]>((a_, b_) => [a_, b_], a, b);

export const zip3 = <A, B, C>(
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
  c: AsyncIterable<C>,
): AsyncIterable<[A, B, C]> => map2<[A, B], C, [A, B, C]>(([a_, b_], c_) => [a_, b_, c_], zip(a, b), c);

export const zip4 = <A, B, C, D>(
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
  c: AsyncIterable<C>,
  d: AsyncIterable<D>,
): AsyncIterable<[A, B, C, D]> =>
  map2<[A, B], [C, D], [A, B, C, D]>(([a_, b_], [c_, d_]) => [a_, b_, c_, d_], zip(a, b), zip(c, d));

export const filter = <A>(fn: (a: A) => Promise<boolean> | boolean, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    for await (let next of iter) {
      if ((await fn(next)) === true) {
        yield next;
      }
    }
  });

export const scan = <A>(fn: (a: A) => Promise<A> | A, init: A): AsyncIterable<A> =>
  create<A>(async function*() {
    let next = init;
    yield next;

    while (true) {
      next = await fn(next);
      yield next;
    }
  });

export const reduce = async <A, B>(
  fn: (prev: A, next: B) => A | Promise<A>,
  acc: A,
  iter: AsyncIterable<B>,
): Promise<A> => {
  let prev = acc;

  for await (let next of iter) {
    prev = await fn(prev, next);
  }

  return prev;
};

export const take = <A>(i: number, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let j = 0;

    for await (let next of iter) {
      yield next;

      if (++j >= i) {
        break;
      }
    }
  });

export const drop = <A>(i: number, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let j = 0;

    for await (let next of iter) {
      if (++j <= i) {
        continue;
      }

      yield next;
    }
  });

export const range = (low: number, high: number): AsyncIterable<number> => {
  if (low > high) {
    return range(high, low);
  }

  return create<number>(async function*() {
    for (let i = low; i < high; i++) {
      yield i;
    }
  });
};

export const cons = <A>(a: A | Promise<A>, xs: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield await a;

    for await (let next of xs) {
      yield next;
    }
  });

export const concat = <A>(a: AsyncIterable<A>, b: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    for await (let next of a) {
      yield next;
    }

    for await (let next of b) {
      yield next;
    }
  });

export const concat3 = <A>(a: AsyncIterable<A>, b: AsyncIterable<A>, c: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* concat(concat(a, b), c);
  });

export const concat4 = <A>(
  a: AsyncIterable<A>,
  b: AsyncIterable<A>,
  c: AsyncIterable<A>,
  d: AsyncIterable<A>,
): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* concat(concat3(a, b, c), d);
  });

export const flatMap = <A, B>(fn: (a: A) => AsyncIterable<B>, iter: AsyncIterable<A>): AsyncIterable<B> =>
  create<B>(async function*() {
    for await (let xs of map(fn, iter)) {
      yield* xs;
    }
  });

export const endless = <A>(iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let isEmpty = true;

    while (true) {
      for await (let value of iter) {
        isEmpty = false;
        yield value;
      }

      if (isEmpty) {
        break;
      }
    }
  });

export const always = <A>(a: A): AsyncIterable<A> =>
  create<A>(async function*() {
    while (true) {
      yield a;
    }
  });

export const reverse = <A>(iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    const cache = [];

    for await (let value of iter) {
      cache.unshift(value);
    }

    yield* cache;
  });

export const cached = <A>(iter: AsyncIterable<A>): AsyncIterable<A> => {
  const cache: A[] = [];
  const iterator = getIterator(iter);
  let done = false;

  return create<A>(async function*() {
    yield* cache;

    if (done === true) {
      return;
    }

    let next: IteratorResult<A>;

    do {
      next = await iterator.next();

      if (next.done === false) {
        const value = next.value;
        cache.push(value);
        yield value;
      }
    } while (next.done === false);

    done = true;
  });
};
