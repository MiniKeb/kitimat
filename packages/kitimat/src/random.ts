export type Seed = { state: number; increment: number };

export const MAX_INT = 2147483647;
export const MIN_INT = -2147483648;

export type Result<A> = {
  value: A;
  nextSeed: Seed;
};

export type Generator<A> = (seed: Seed) => Promise<Result<A>>;

const next = (seed: Seed): Seed => {
  const { state, increment } = seed;
  // The magic constant is from Numerical Recipes and is inlined for perf.
  const next = (state * 1664525 + increment) >>> 0;
  return { state: next, increment };
};

const peel = (seed: Seed): number => {
  const { state } = seed;
  const word = ((state >>> ((state >>> 28) + 4)) ^ state) * 277803737;
  return ((word >>> 22) | word) >>> 0;
};

export const initialSeed = (x: number): Seed => {
  // The magic constant is from Numerical Recipes and is inlined for perf.
  const seed: Seed = next({ state: 0, increment: 1013904223 });
  const state2 = seed.state + (x >>> 0);
  return next({ state: state2, increment: seed.increment });
};

export const flatMap = <A, B>(fn: (a: A) => Generator<B>, a: Generator<A>): Generator<B> => async seed => {
  const { value, nextSeed } = await a(seed);
  const nextGen = fn(value);
  return nextGen(nextSeed);
};

export const map = <A, B>(fn: (a: A) => B | Promise<B>, a: Generator<A>): Generator<B> => async seed => {
  const { value, nextSeed } = await a(seed);
  return { value: await fn(value), nextSeed };
};

export const map2 = <A, B, C>(
  fn: (a: A, b: B) => C | Promise<C>,
  a: Generator<A>,
  b: Generator<B>,
): Generator<C> => async seed => {
  const { value: valA, nextSeed: seedA } = await a(seed);
  const { value: valB, nextSeed: seedB } = await b(seedA);
  return { value: await fn(valA, valB), nextSeed: seedB };
};

export const frequency = <A>(pairs: [number, Generator<A>][]): Generator<A> => {
  let total = 0;
  let totaledWeights: [number, Generator<A>][] = [];
  let i = -1;

  for (let pair of pairs) {
    total = pair[0] + total;
    const next: [number, Generator<A>] = [total, pair[1]];
    totaledWeights = [...totaledWeights, next];
  }

  const generator = integer(0, total - 1);

  return async seed => {
    const { value: pick, nextSeed } = await generator(seed);

    for (let pair of totaledWeights) {
      if (pair[0] > pick) {
        return pair[1](nextSeed);
      }
    }

    // impossible
    return totaledWeights[0][1](nextSeed);
  };
};

export const array = <A>(maxLen: number, gen: Generator<A>): Generator<A[]> => {
  const sizeGenerator = integer(0, maxLen);

  return async seed => {
    let i = -1;
    let xs: A[] = [];

    let { value: size, nextSeed } = await sizeGenerator(seed);

    while (++i < size) {
      const result = await gen(nextSeed);
      nextSeed = result.nextSeed;
      xs = xs.concat(result.value);
    }

    return { value: xs, nextSeed };
  };
};

export const constant = <A>(value: A): Generator<A> => async nextSeed => {
  return { value, nextSeed };
};

export const filter = <A>(fn: (a: A) => boolean, a: Generator<A>): Generator<A> => seed_ => {
  const tryNext: Generator<A> = async seed => {
    const { value, nextSeed } = await a(seed);
    return fn(value) === true ? { value, nextSeed } : tryNext(nextSeed);
  };

  return tryNext(seed_);
};

export const filterMap = <A>(fn: (a: A) => A | void, a: Generator<A>): Generator<A> => seed_ => {
  const tryNext: Generator<A> = async seed => {
    const { value, nextSeed } = await a(seed);
    const nextValue = fn(value);

    return nextValue !== undefined ? { value: nextValue, nextSeed } : tryNext(nextSeed);
  };

  return tryNext(seed_);
};

export const integer = (a: number, b: number): Generator<number> => async seed => {
  let min: number;
  let max: number;

  if (a < b) {
    min = a;
    max = b;
  } else {
    min = b;
    max = a;
  }

  const range = max - min + 1;

  // this is the same as asking, "is range a power of 2?""
  // we do this because the power of 2 path is 150% faster than if it isn't
  if (((range - 1) & range) === 0) {
    const value = ((peel(seed) & (range - 1)) >>> 0) + min;
    return { value, nextSeed: next(seed) };
  }

  const threshold = ((-range >>> 0) % range) >>> 0;

  const accountForBias = (seed: Seed): Result<number> => {
    const x = peel(seed);
    const nextSeed = next(seed);

    if (x < threshold) {
      return accountForBias(nextSeed);
    } else {
      return { value: x % range + min, nextSeed };
    }
  };

  return accountForBias(seed);
};

export const float = (min: number, max: number): Generator<number> => async seed => {
  const bit53 = 9007199254740992;
  const bit27 = 134217728;
  const nextSeed = next(seed);
  const n0 = peel(seed);
  const n1 = peel(nextSeed);

  const high = n0 & 0x03ffffff;
  const low = n1 & 0x07ffffff;

  const val = (high * bit27 + low) / bit53;
  const range = Math.abs(max - min);
  const scaled = val * range + min;

  return { value: scaled, nextSeed: next(nextSeed) };
};

export const boolean: Generator<boolean> = map(i => i === 1, integer(0, 1));

export const character = (minCharacter: number, maxCharacter: number): Generator<string> =>
  map(i => String.fromCharCode(i), integer(minCharacter, maxCharacter));

export const asciiCharacter = character(32, 126);

export const string = (minSize: number, maxSize: number): Generator<string> =>
  map(chars => chars.join(''), flatMap(size => array(size, character(32, 255)), integer(minSize, maxSize)));

export const asciiString = (minSize: number, maxSize: number): Generator<string> =>
  map(chars => chars.join(''), flatMap(size => array(size, asciiCharacter), integer(minSize, maxSize)));

export const sample = <A>(constants: A[]): Generator<A> => {
  const pairs = constants.map<[number, Generator<A>]>(c => [1, constant(c)]);
  return frequency(pairs);
};

export const whitespaceCharacter = sample([' ', '\t', '\n']);

export const whitespace: Generator<string> = map(
  chars => chars.join(''),
  flatMap(size => array(size, whitespaceCharacter), integer(1, 10)),
);
