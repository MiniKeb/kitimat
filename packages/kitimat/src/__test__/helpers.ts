import * as Rose from '../rosetree';
import * as Random from '../random';
import * as Iter from '../iterable';
import { sample as endlessSample } from '../sample';

export const sample = <A>(a: Random.Generator<A>, size = 10, seed?: Random.Seed): AsyncIterable<A> =>
  Iter.take(size, endlessSample<A>(a, seed));
