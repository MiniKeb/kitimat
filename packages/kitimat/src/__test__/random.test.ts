import { sample } from './helpers';
import * as Random from '../random';
import * as Iter from '../iterable';

const DEVIATION = 0.05;

const staticSeed = Random.initialSeed(1514568898760);
const randomSeed = Random.initialSeed(Date.now());

describe('unit', () => {
  test('sample/integer', async () => {
    const integer = Random.integer(-1e9, 1e9);

    const result = await Iter.toArray(sample(integer, 20, staticSeed));
    expect(result).toMatchSnapshot();
    expect(result.length).toEqual(20);
  });

  test('float', async () => {
    const float = Random.float(-1e9, 1e9);
    expect(await Iter.toArray(sample(float, 20, staticSeed))).toMatchSnapshot();
  });

  test('boolean', async () => {
    const boolean = Random.boolean;
    expect(await Iter.toArray(sample(boolean, 20, staticSeed))).toMatchSnapshot();
  });

  test('string', async () => {
    const string = Random.string(0, 10);
    expect(await Iter.toArray(sample(string, 20, staticSeed))).toMatchSnapshot();
  });

  test('array<integer>', async () => {
    const array = Random.array(10, Random.integer(-100, 100));
    expect(await Iter.toArray(sample(array, 20, staticSeed))).toMatchSnapshot();
  });

  test('filter', async () => {
    const integer = Random.integer(-100, 100);
    const integerArray = Random.array(100, integer);
    const gtZero = Random.filter(x => x > 0, integer);
    const gtZeroArray = Random.array(100, gtZero);

    const { value: ints, nextSeed: intLastSeed } = await integerArray(staticSeed);
    const { value: posInts, nextSeed: posIntLastSeed } = await gtZeroArray(staticSeed);

    expect(ints.some(x => x < 0)).toBe(true);
    expect(posInts.every(x => x > 0)).toBe(true);
    expect(ints).not.toEqual(posInts);
    expect(intLastSeed).not.toEqual(posIntLastSeed);
  });

  test('weighted', async () => {
    const generator = Random.frequency<boolean | string>([
      [1, Random.constant(true)],
      [2, Random.constant(false)],
      [6, Random.string(0, 40)],
    ]);

    expect(await Iter.toArray(sample(generator, 20, staticSeed))).toMatchSnapshot();
  });

  test('flatMap', async () => {
    const integer = Random.integer(-100, 100);
    const boolean = Random.flatMap(i => (i > 50 ? Random.constant(true) : Random.constant(false)), integer);
    const result = await Iter.toArray(sample(boolean, 20, staticSeed));

    expect(result).toMatchSnapshot();
    expect(result.every(x => typeof x === 'boolean')).toBe(true);
  });
});

describe('property', () => {
  test('integer where min and max are the same', async () => {
    const integer = Random.integer(10, 10);
    const { value } = await integer(randomSeed);
    expect(value).toEqual(10);
  });

  test('integer is within bounds', async () => {
    const min = -10;
    const max = 10;
    const integer = Random.integer(min, max);
    const { value } = await integer(randomSeed);
    expect(value).toBeLessThanOrEqual(max);
    expect(value).toBeGreaterThanOrEqual(min);
  });

  test('float is within bounds', async () => {
    const min = -1;
    const max = 1;
    const float = Random.float(min, max);
    const { value } = await float(randomSeed);
    expect(value).toBeLessThanOrEqual(max);
    expect(value).toBeGreaterThanOrEqual(min);
  });

  test('average integer is approximately (min + max) / 2', async () => {
    const runs = 1e4;
    let total = 0;
    const integer = Random.integer(-100, 100);
    const expected = 0;

    for await (let n of sample(integer, runs)) {
      total += n;
    }

    const average = total / runs;

    expect(average).toBeLessThan(expected + runs * DEVIATION);
    expect(average).toBeGreaterThan(expected - runs * DEVIATION);
  });

  test('average float is approximately (min + max) / 2', async () => {
    const runs = 1e4;
    let total = 0;
    const integer = Random.float(0, 200);
    const expected = 100;

    for await (let n of sample(integer, runs)) {
      total += n;
    }

    const average = total / runs;

    expect(average).toBeLessThan(expected + runs * DEVIATION);
    expect(average).toBeGreaterThan(expected - runs * DEVIATION);
  });

  test('same number of trues as falses', async () => {
    let average = 0;
    const runs = 1e4;

    for await (let n of sample(Random.boolean, runs)) {
      average += n ? 1 : -1;
    }

    expect(average).toBeLessThan(runs * DEVIATION);
    expect(average).toBeGreaterThan(-runs * DEVIATION);
  });

  test('weighted generators should return weighted results', async () => {
    const counts: { [key in 'true' | 'false' | 'undefined']: number } = {
      true: 0,
      false: 0,
      undefined: 0,
    };

    const runs = 1e4;
    const tWeight = 1;
    const fWeight = 4;
    const uWeight = 2;

    const generator = Random.frequency<boolean | void>([
      [tWeight, Random.constant(true)],
      [fWeight, Random.constant(false)],
      [uWeight, Random.constant(undefined)],
    ]);

    for await (let n of sample(generator, runs)) {
      counts[JSON.stringify(n) as 'true' | 'false' | 'undefined'] += 1;
    }

    const tExpected = runs * (tWeight / 7);
    const fExpected = runs * (fWeight / 7);
    const uExpected = runs * (uWeight / 7);

    expect(counts.true).toBeLessThan(tExpected + runs * DEVIATION);
    expect(counts.true).toBeGreaterThan(tExpected - runs * DEVIATION);
    expect(counts.false).toBeLessThan(fExpected + runs * DEVIATION);
    expect(counts.false).toBeGreaterThan(fExpected - runs * DEVIATION);
    expect(counts.undefined).toBeLessThan(uExpected + runs * DEVIATION);
    expect(counts.undefined).toBeGreaterThan(uExpected - runs * DEVIATION);
  });
});
