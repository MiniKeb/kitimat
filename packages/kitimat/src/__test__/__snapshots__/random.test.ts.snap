// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit array<integer> 1`] = `
Array [
  Array [],
  Array [
    -18,
  ],
  Array [],
  Array [
    56,
    9,
    -50,
    43,
    38,
    22,
    -44,
    47,
    -96,
    64,
  ],
  Array [
    -22,
    -70,
    -68,
    -15,
    -9,
  ],
  Array [
    -8,
    77,
    93,
  ],
  Array [
    70,
    -58,
    -7,
    93,
    -15,
    -20,
    -93,
    48,
    75,
  ],
  Array [
    -68,
    -14,
    -91,
    34,
    -19,
  ],
  Array [
    -38,
    66,
  ],
  Array [
    -55,
    48,
    32,
    67,
    75,
  ],
  Array [
    -55,
    11,
    -98,
    71,
    -65,
    55,
    -50,
    -80,
  ],
  Array [
    5,
    61,
    9,
  ],
  Array [
    55,
    -97,
    98,
    -16,
  ],
  Array [
    78,
    54,
    -46,
    65,
    -7,
    86,
  ],
  Array [
    -78,
    0,
    -46,
    78,
    39,
    90,
  ],
  Array [
    27,
    14,
    49,
    -33,
    4,
    -6,
    61,
    91,
  ],
  Array [
    -14,
    17,
    -53,
    93,
  ],
  Array [
    -99,
    50,
    -69,
    -64,
    -100,
    -68,
    -79,
  ],
  Array [],
  Array [
    -50,
    0,
    -98,
  ],
]
`;

exports[`unit boolean 1`] = `
Array [
  false,
  false,
  true,
  true,
  true,
  false,
  true,
  true,
  false,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  true,
  true,
  false,
]
`;

exports[`unit flatMap 1`] = `
Array [
  true,
  false,
  false,
  false,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  true,
  false,
  false,
  false,
  false,
]
`;

exports[`unit float 1`] = `
Array [
  337341157.7788298,
  131103315.70345783,
  -337528066.4467977,
  677046229.7624793,
  564574149.1771181,
  235102469.6342833,
  981104846.5482476,
  -883028001.6414496,
  -474954005.57094246,
  735809255.6201799,
  -560800444.1412815,
  864524318.5914454,
  -278001076.74858236,
  939590432.5767307,
  -269151523.55651987,
  704159657.843962,
  621960071.0778644,
  252468888.8512342,
  -200379170.24436247,
  -885592448.4652789,
]
`;

exports[`unit sample/integer 1`] = `
Array [
  -794376712,
  615757277,
  863158776,
  112899337,
  632841605,
  267863534,
  666885069,
  164663797,
  -853860902,
  -987222598,
  -623012519,
  735329785,
  623244638,
  -603089670,
  -311720513,
  -199414594,
  37516499,
  223655267,
  -713897477,
  561311604,
]
`;

exports[`unit string 1`] = `
Array [
  "",
  ".Mõ",
  "z",
  "Àþ",
  "4",
  "î.",
  "ß",
  "0]«lý",
  "&ø",
  "äj8'ç¼ã",
  "",
  "¦Ô®¼",
  "",
  "",
  "ÿZh¤",
  "]>YF",
  "d",
  "6îµ¨ùÖ×",
  "",
  "O¼¼ºan°",
]
`;

exports[`unit weighted 1`] = `
Array [
  "©.",
  false,
  "Ùz_ÀþÃ]4ãôî.",
  "¥0]«lýº&øÀÍ",
  "'ç¼ã¶MWW¦",
  "¼",
  false,
  false,
  "UÇÿZh¤çx]>YF§dc6îµ¨ùÖ×Ò}Jß",
  false,
  "º",
  "xýÅI",
  false,
  true,
  "ìz$hP=ª2äÝ",
  "<M?XvP$",
  true,
  "±;I#ì§vqÄ·G³ )½öY\\"¡W}ô6",
  false,
  "ôF|ÿÿq±;",
]
`;
